projects/RabbitMQ.Client/client/TaskExtensions.cs

```diff
@@ -59,21 +59,29 @@ public static Task TimeoutAfter(this Task task, TimeSpan timeout)
 
             return DoTimeoutAfter(task, timeout);
 
+            // https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#using-a-timeout
             static async Task DoTimeoutAfter(Task task, TimeSpan timeout)
             {
-                if (task == await Task.WhenAny(task, Task.Delay(timeout)).ConfigureAwait(false))
+                using (var cts = new CancellationTokenSource())
                 {
-                    await task.ConfigureAwait(false);
-                }
-                else
-                {
-                    Task supressErrorTask = task.ContinueWith(
-                        continuationAction: continuation,
-                        state: null,
-                        cancellationToken: CancellationToken.None,
-                        continuationOptions: s_tco,
-                        scheduler: TaskScheduler.Default);
-                    throw new TimeoutException();
+                    Task delayTask = Task.Delay(timeout, cts.Token);
+                    Task resultTask = await Task.WhenAny(task, delayTask).ConfigureAwait(false);
+                    if (resultTask == delayTask)
+                    {
+                        Task supressErrorTask = task.ContinueWith(
+                            continuationAction: continuation,
+                            state: null,
+                            cancellationToken: CancellationToken.None,
+                            continuationOptions: s_tco,
+                            scheduler: TaskScheduler.Default);
+                        throw new TimeoutException();
```

Comment:
`OperationCanceledException` (or, more specifically for tasks, `TaskCancelledException`) and `TimeoutException` are two different things.

And what we have here is a timeout, not a cancellation.

Am I missing something here?

Reply:
There are multiple examples. For example the implementation in the guidance also uses OperationCanceledException see https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#using-a-timeout

That being said, based on my experience, the canonical example for any sort of cancellation of tasks (timeout, token based or token with timeout (cancelafter)) was always either `OperationCanceledException` or more specifically `TaskCancelledException`. The Azure SDK for example, favors the derived `TaskCancelledException`.

Reply:
My bad. That's what [Task.WaitAsync](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitasync) does.

That's not exactly an operation cancelled, because it does nothing for the operation. Just stops awaiting for it.

But, being internal, I'd favor using a `CancellationTokenSource` with a timeout, because there should be a cancellation token in every asynchronous operation.

Reply:
Just learned something from this discussion. Apparently the behavior depends on the overload. For example this overload does throw a Timeout Exception https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitasync?view=net-8.0#system-threading-tasks-task-waitasync(system-timespan-system-timeprovider)

Reply:
That was my initial understanding. But that is a very poor piece of documentation. Mentioned the outcome of the operation only in one of the parameters is odd, to say the least.

Reply:
> But, being internal, I'd favor using a CancellationTokenSource with a timeout, because there should be a cancellation token in every asynchronous operation.

That's the best option. There are only a couple places where these task extensions are even used, and one of them is for timing out a socket connection, which does support cancellation.

Reply:
I'm going to start work on #1420, which should remove the use of `TimeoutAfter` in the place where users of this library are most likely to encounter a timeout (socket connection - https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs#L333-L361)

Reply:
I have linked to this discussion here: https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1425